### PR TITLE
Host-side API for view-owned widgets

### PR DESCRIPTION
**Pull request by squarejesse**
This introduces a new mutable property, windowInsets, that is propagated to the composition where it can be consumed.

We put the burden on the user of RedwoodView to set this value to something non-zero if they have insets to be consumed. I've added an accelerator on Android for this,

```kotlin
RedwoodLayout.consumeWindowInsets(mask: Int)
```

### CODE REVIEW

File: redwood-widget/src/androidMain/kotlin/app/cash/redwood/widget/RedwoodLayout.kt [hunk 1]
```diff
@@ -112,7 +118,7 @@ public open class RedwoodLayout(
 
   private fun computeUiConfiguration(
     config: Configuration = context.resources.configuration,
-    insets: Insets = rootWindowInsetsCompat.safeDrawing,
```
Comment (squarejesse) [inline redwood-widget/src/androidMain/kotlin/app/cash/redwood/widget/RedwoodLayout.kt:115]: # ☠️

This is the core problem of the old API. If we end up calling this function for a configuration change, it forgets what the *view-local* insets were and reverts to the *window* insets.

File: redwood-runtime/src/commonMain/kotlin/app/cash/redwood/ui/UiConfiguration.kt [hunk 2]
```diff
@@ -25,13 +25,23 @@ public class UiConfiguration(
...
+
+  /**
+   * The insets of the viewport that the composition is responsible for consuming.
+   *
+   * This may be zero if the host view isn't attached to a view hierarchy and therefore doesn't
+   * know its insets.
+   *
+   * See https://developer.android.com/develop/ui/views/layout/edge-to-edge
+   */
+  public val viewInsets: Margin = Margin.Zero,
```
Comment (squarejesse) [inline redwood-runtime/src/commonMain/kotlin/app/cash/redwood/ui/UiConfiguration.kt:43]: Unfortunately `safeAreaInsets` are potentially non-zero on already-shipping clients, but they don’t necessarily apply to the view we want them to.

Instead I’m creating another property that’s my attempt at fixing the window-local vs. view-local problem.

File: redwood-treehouse-host/src/androidUnitTest/kotlin/app/cash/redwood/treehouse/TreehouseLayoutTest.kt [hunk 3]
```diff
@@ -112,7 +115,9 @@ class TreehouseLayoutTest {
           bottom = insets.bottom.toDp(),
         )
       }
-      assertThat(awaitItem().safeAreaInsets).isEqualTo(expectedInsets)
+      val value2 = awaitItem()
+      assertThat(value2.safeAreaInsets).isEqualTo(Margin.Zero)
+      assertThat(value2.viewInsets).isEqualTo(expectedInsets)
```
Comment (JakeWharton) [inline redwood-treehouse-host/src/androidUnitTest/kotlin/app/cash/redwood/treehouse/TreehouseLayoutTest.kt:120]: Change log item for this behavior change? Probably want to give a heads-up to internal users as well, if they still exist, that is.

**Review by JakeWharton — APPROVED** [review]

File: redwood-treehouse-host/src/androidUnitTest/kotlin/app/cash/redwood/treehouse/TreehouseLayoutTest.kt (hunk 3 above)
Comment (squarejesse) [reply redwood-treehouse-host/src/androidUnitTest/kotlin/app/cash/redwood/treehouse/TreehouseLayoutTest.kt:120]: Yep, lemme do.

**Pull request merged**